### PR TITLE
Load local git name/email for cargo init

### LINKED ISSUE #3920: Load author name and email from innermost git config
It appears that `cargo init` always uses the name and email configured in ~/.gitconfig. It would be more useful (especially when working with workspaces with many crates) if `cargo init` produced the same output as `git config user.name` and `git config user.email` which can be configured per repo.

### PR DESCRIPTION
**Pull request by huntiep**
Fixes #3920

### CODE REVIEW

**Comment by huntiep** [comment]
Git 2.13 added an `includeIf` config option that makes working with multiple git identities easier. I thought it would be worth testing; so I installed git 2.13 and tested this with Cargo, and it seems to work properly.

**Comment by alexcrichton** [comment]
Thanks for the PR! I wonder, could a test be added for this as well?

File: src/cargo/ops/cargo_new.rs [hunk 1]
```diff
@@ -514,7 +515,12 @@ fn get_environment_variable(variables: &[&str] ) -> Option<String>{
 }
 
 fn discover_author() -> CargoResult<(String, Option<String>)> {
-    let git_config = GitConfig::open_default().ok();
+    let cwd = env::current_dir()?;
+    let git_config = if let Ok(repo) = GitRepository::discover(&cwd) {
+        repo.config().ok()
```
Comment (huntiep) [inline src/cargo/ops/cargo_new.rs]: Isn't that being done in the else block, or did you mean something else?

**Comment by huntiep** [comment]
Yeah, how do you want me to do this? I think that we already need a git repository to test this, should I create a TempDir and initialize a git repository there?

File: src/cargo/ops/cargo_new.rs (hunk 1 above)
Comment (alexcrichton) [inline src/cargo/ops/cargo_new.rs]: Oh I just means that if `repo.config().ok()` is none, we fall back to `open_default`. So we'd fall back either if `discover` failed or if `config` failed on the opened repository

**Comment by alexcrichton** [comment]
Oh there's a number of other tests in `tests/*.rs` which work with git repositories, should be some examples to draw from.

**Comment by alexcrichton** [comment]
ping @huntiep, any updates here on the tests?

**Comment by huntiep** [comment]
I won't have access to a devenv until tomorrow. I'll work on it then.

**Comment by huntiep** [comment]
I was wrong about cargo working properly with git 2.13's includeIf. I'm not sure what gave me the idea that it worked. It looks like support is dependent on libgit2. I'll open in issue in git2-rs about it.

**Comment by alexcrichton** [comment]
@bors: r+

Thanks!

**Pull request merged**